src/main/scala/org/broadinstitute/hail/Utils.scala

```diff
@@ -324,6 +324,53 @@ class RichSparkContext(val sc: SparkContext) extends AnyVal {
       .map(l => WithContext(l, TextContext(l, file, None)))
 }
 
+class HeadRDD[T](prev: RDD[T], num: Int, exact: Boolean)(implicit tct: ClassTag[T]) extends RDD[T](prev) {
+
+  private var nLastPart = 0
+  private var lastPart = 0
+
+  def getPartitions: Array[Partition] = {
+    val buf = new mutable.ArrayBuffer[Int]
+
+    buf ++= prev.sparkContext.runJob(prev, (it: Iterator[T]) => it.length, Seq(0))
+    if (exact || buf.length == 0) {
+      while (buf.sum < num && buf.length < prev.partitions.length) {
+        buf ++= prev.sparkContext.runJob(prev, (it: Iterator[T]) => it.length,
+          buf.length to
+            buf.length +
+              (if (buf.sum > 0) math.min(((num - buf.sum) / (buf.sum / buf.length).toDouble).ceil.toInt, prev.partitions.length)
+              else 1)
```

Comment:
If the first bunch of partitions are empty, or very small compared to what's to come, this isn't going to perform very well.  A few ideas:

 - I'd start with 4 partitions, not 1.  Most systems we run on will have at least 4 executors (even laptops), so it should essentially be free.

 - You might increase exponentially by 4s until you reach some fraction of the target (10%?)  If you ask for 10k and there is 1 element in the first partition, you're going to compute the entire thing even if the second partition has >10k entries.

 - Use the >10% as an estimate in no-exact mode, and a first guess in exact mode.

What do you think?